File: agent/collectors/printers.py
import platform
import wmi
import pythoncom
import win32evtlog
import xml.etree.ElementTree as ET
from core.config import load_config, save_config

def get_printer_metrics():
    """Scan and sum print logs since last check"""
    config = load_config()
    page_counts = config.get("printer_page_counts", {})
    last_id = config.get("last_processed_event_record_id", 0)
    
    log_path = "Microsoft-Windows-PrintService/Operational"
    query_str = f"*[System[EventID=307 and EventRecordID > {last_id}]]"
    
    max_record_id = last_id
    
    try:
        handle = win32evtlog.EvtQuery(log_path, win32evtlog.EvtQueryChannelPath, query_str)
        while True:
            events = win32evtlog.EvtNext(handle, 100)
            if not events:
                break
                
            for event in events:
                try:
                    xml_str = win32evtlog.EvtRender(event, win32evtlog.EvtRenderEventXml)
                    root = ET.fromstring(xml_str)
                    
                    ns_sys = {'ns': 'http://schemas.microsoft.com/win/2004/08/events/event'}
                    rec_elem = root.find(".//ns:EventRecordID", ns_sys)
                    if rec_elem is not None and rec_elem.text:
                        rec_id = int(rec_elem.text)
                        if rec_id > max_record_id:
                            max_record_id = rec_id
                            
                    ns_user = {'ns': 'http://manifests.microsoft.com/win/2005/08/windows/printing/spooler/core/events'}
                    p5_elem = root.find(".//ns:Param5", ns_user)
                    p8_elem = root.find(".//ns:Param8", ns_user)
                    
                    if p5_elem is not None and p8_elem is not None and p5_elem.text and p8_elem.text:
                        printer_name = p5_elem.text.strip()
                        pages = int(p8_elem.text)
                        
                        if not (printer_name.startswith('{') and printer_name.endswith('}')):
                            page_counts[printer_name] = page_counts.get(printer_name, 0) + pages
                except Exception as e:
                    pass
    except Exception as e:
        print(f"[PrinterLog] EvtQuery failed: {e}")
        
    config["printer_page_counts"] = page_counts
    config["last_processed_event_record_id"] = max_record_id
    save_config(config)
    return page_counts

def get_printers_info():
    """Scan connected printers using WMI and get print log metrics"""
    pythoncom.CoInitialize()
    printers_info = []
    if platform.system() != "Windows":
        return printers_info
        
    try:
        c = wmi.WMI()
        page_counts = get_printer_metrics()
        drivers_info = {}
        
        for driver in c.Win32_PrinterDriver():
            name_key = driver.Name.split(",")[0] if driver.Name else ""
            drivers_info[name_key] = {
                "driver_version": getattr(driver, "DriverVersion", "Unknown"),
                "manufacturer": getattr(driver, "Manufacturer", "Unknown")
            }

        seen_printers = set()
        for printer in c.Win32_Printer():
            name = getattr(printer, "Name", "Unknown")
            if name in seen_printers:
                continue
            seen_printers.add(name)
            
            driver_name = getattr(printer, "DriverName", "")
            driver_details = drivers_info.get(driver_name, {"driver_version": "Unknown", "manufacturer": "Unknown"})
            
            total_pages = page_counts.get(name, 0)
            
            printers_info.append({
                "name": name,
                "model": getattr(printer, "DeviceID", "Unknown"),
                "port": getattr(printer, "PortName", "N/A"),
                "network": getattr(printer, "Network", False),
                "manufacturer": driver_details["manufacturer"],
                "driver_version": driver_details["driver_version"],
                "total_pages_printed": total_pages
            })
    except Exception as e:
        print(f"[Printer] WMI scan error: {e}")
        
    return printers_info

File: go_agent/collectors/printers.go
package collectors

import (
	"encoding/xml"
	"fmt"
	"os/exec"
	"strings"
	"strconv"

	"go_agent/config"
	"go_agent/logger"
)

type PrinterInfo struct {
	Name              string `json:"name"`
	Model             string `json:"model"`
	Port              string `json:"port"`
	IsNetwork         bool   `json:"network"`
	Manufacturer      string `json:"manufacturer"`
	DriverVersion     string `json:"driver_version"`
	TotalPagesPrinted int    `json:"total_pages_printed"`
}

// ---- Event Log XML parsing ----

type eventXML struct {
	System struct {
		EventRecordID int64 `xml:"EventRecordID"`
	} `xml:"System"`
	EventData struct {
		Data []struct {
			Name  string `xml:"Name,attr"`
			Value string `xml:",chardata"`
		} `xml:"Data"`
	} `xml:"UserData>PrintJobDetailedInfo,omitempty"`
}

// getPrintEventsXML uses wevtutil to query print job events (ID 307)
func getPrintEventsXML(lastRecordID int64) ([]eventXML, int64) {
	// wevtutil qe "Microsoft-Windows-PrintService/Operational" /q:"*[System[EventID=307]]" /f:xml /rd:false /e:root
	// We filter by record ID on our side (wevtutil doesn't easily support EventRecordID filter)
	args := []string{
		"qe",
		"Microsoft-Windows-PrintService/Operational",
		`/q:*[System[EventID=307]]`,
		"/f:xml",
		"/rd:false",
		"/e:root",
	}
	out, err := exec.Command("wevtutil", args...).Output()
	if err != nil {
		logger.Error("Printers", fmt.Sprintf("wevtutil failed: %v", err))
		return nil, lastRecordID
	}

	// Wrap in root element for xml parsing
	wrapped := "<wrapper>" + string(out) + "</wrapper>"
	type wrapper struct {
		Events []eventXML `xml:"Event"`
	}
	var w wrapper
	if err := xml.Unmarshal([]byte(wrapped), &w); err != nil {
		logger.Error("Printers", fmt.Sprintf("XML parse error: %v", err))
		return nil, lastRecordID
	}

	maxID := lastRecordID
	var filtered []eventXML
	for _, ev := range w.Events {
		if ev.System.EventRecordID > lastRecordID {
			filtered = append(filtered, ev)
			if ev.System.EventRecordID > maxID {
				maxID = ev.System.EventRecordID
			}
		}
	}
	return filtered, maxID
}

// getPrinterMetrics reads print events and accumulates page counts
func getPrinterMetrics() map[string]int {
	cfg := config.Load()
	pageCounts := cfg.PrinterPageCounts
	if pageCounts == nil {
		pageCounts = map[string]int{}
	}

	events, maxID := getPrintEventsXML(cfg.LastProcessedEventRecordID)
	for _, ev := range events {
		// Map EventData fields: Param5 = printer name, Param8 = pages
		dataMap := map[string]string{}
		for _, d := range ev.EventData.Data {
			dataMap[d.Name] = d.Value
		}
		printerName := strings.TrimSpace(dataMap["Param5"])
		pagesStr := strings.TrimSpace(dataMap["Param8"])
		if printerName == "" || strings.HasPrefix(printerName, "{") {
			continue
		}
		pages, err := strconv.Atoi(pagesStr)
		if err != nil || pages <= 0 {
			continue
		}
		pageCounts[printerName] += pages
		logger.Info("Printers", fmt.Sprintf("EventLog: +%d pages → %s", pages, printerName))
	}

	cfg.PrinterPageCounts = pageCounts
	cfg.LastProcessedEventRecordID = maxID
	_ = config.Save(cfg)
	return pageCounts
}

// GetPrintersInfo scans installed printers via WMI and combines with event log metrics
func GetPrintersInfo() []PrinterInfo {
	pageCounts := getPrinterMetrics()

	wmi, service, err := wmiConnect()
	if err != nil {
		logger.Error("Printers", fmt.Sprintf("WMI connect error: %v", err))
		return nil
	}
	defer wmi.Release()
	defer service.Release()
	defer oleUninitialize()

	// Collect driver info
	driverMap := map[string]struct {
		version      string
		manufacturer string
	}{}
	dItems, dCleanup := wmiQueryItems(service,
		"SELECT Name,DriverVersion,Manufacturer FROM Win32_PrinterDriver")
	defer dCleanup()
	for _, it := range dItems {
		name := strProp(it, "Name")
		// Driver names look like "DriverName,3,Windows x64" – take first part
		parts := strings.SplitN(name, ",", 2)
		key := strings.TrimSpace(parts[0])
		driverMap[key] = struct {
			version      string
			manufacturer string
		}{
			version:      strProp(it, "DriverVersion"),
			manufacturer: strProp(it, "Manufacturer"),
		}
	}

	// Collect printers
	pItems, pCleanup := wmiQueryItems(service,
		"SELECT Name,DeviceID,PortName,Network,DriverName FROM Win32_Printer")
	defer pCleanup()

	seen := map[string]bool{}
	result := []PrinterInfo{}
	for _, it := range pItems {
		name := strProp(it, "Name")
		if name == "" || seen[name] {
			continue
		}
		seen[name] = true

		driverName := strProp(it, "DriverName")
		di := driverMap[driverName]

		networkStr := strProp(it, "Network")
		isNetwork := strings.EqualFold(networkStr, "true") || networkStr == "1"

		result = append(result, PrinterInfo{
			Name:              name,
			Model:             strProp(it, "DeviceID"),
			Port:              strProp(it, "PortName"),
			IsNetwork:         isNetwork,
			Manufacturer:      di.manufacturer,
			DriverVersion:     di.version,
			TotalPagesPrinted: pageCounts[name],
		})
	}
	return result
}

// oleUninitialize is a convenience passthrough so we can defer it
func oleUninitialize() {
	// go-ole auto-tracks per-goroutine init/uninit via CoInitializeEx
	// We call our shared connector's uninit here
	// (no-op here because wmiConnect does it inline in this package)
}
